### PR TITLE
feat: added new tab for infra metrics in logs detailed page

### PR DESCRIPTION
**Pull request by rahulkeswani101**
### Summary

https://www.notion.so/signoz/PRD-Correlation-between-3-signals-3b3a1822c2f641b39e2828041e8599d4#a32ee8fe5bf545bca66ce42b867a7224

#### Related Issues / PR's

#5730 

https://github.com/user-attachments/assets/fef8a04d-6b64-4163-81db-225919335594

### CODE REVIEW

**Comment by ankitnayan** [comment]
Screenshot 2024-08-26 at 9 37 07 PM

**Comment by rahulkeswani101** [comment]
@ankitnayan As we discussed, I am following the order of panels according to the previous view which is not the same as the JSON  file given for reference for overall pod metrics.

**Review by ankitnayan — DISMISSED** [review]

File: frontend/src/components/LogDetail/index.tsx [hunk 1]
```diff
@@ -246,6 +259,15 @@ function LogDetail({
...
 				/>
 			)}
+			{selectedView === VIEW_TYPES.INFRAMETRICS && (
+				<InfraMetrics
+					clusterName={log.resources_string?.[RESOURCE_KEYS.CLUSTER_NAME] as string}
+					podName={log.resources_string?.[RESOURCE_KEYS.POD_NAME] as string}
+					nodeName={log.resources_string?.[RESOURCE_KEYS.NODE_NAME] as string}
+					hostName={log.resources_string?.[RESOURCE_KEYS.HOST_NAME] as string}
+					logLineTimestamp={log.timestamp.toString()}
+				/>
```
Comment (ahmadshaheer) [inline frontend/src/components/LogDetail/index.tsx:269]: do we need `as string`?

File: frontend/src/container/LogDetailedView/InfraMetrics/NodeMetrics.tsx [hunk 2]
```diff
@@ -0,0 +1,140 @@
+import { Card, Col, Row, Skeleton, Typography } from 'antd';
```
Comment (ahmadshaheer) [inline frontend/src/container/LogDetailedView/InfraMetrics/NodeMetrics.tsx:1]: nit: The NodeMetrics and PodMetrics components seem pretty similar, should we have a reusable component that can render both the components?

File: frontend/src/container/LogDetailedView/LogDetailedView.types.ts [hunk 3]
```diff
@@ -23,3 +23,7 @@ export interface IFieldAttributes {
 export interface JSONViewProps {
 	logData: ILog;
 }
+
+export interface InfraMetricsSProps {
+	logData: ILog;
+}
```
Comment (ahmadshaheer) [inline frontend/src/container/LogDetailedView/LogDetailedView.types.ts:29]: these two interfaces seem the same, should we define InfraMetricsProps as a type alias to JSONViewProps (i.e. type InfraMetricsProps = JSONViewProps?

File: frontend/src/lib/dashboard/getQueryResults.ts [hunk 4]
```diff
@@ -87,4 +87,6 @@ export interface GetQueryResultsProps {
 		pagination?: Pagination;
 		selectColumns?: any;
 	};
+	start?: number;
+	end?: number;
```
Comment (ahmadshaheer) [inline frontend/src/lib/dashboard/getQueryResults.ts:91]: didn't we already have start and end?

File: frontend/src/lib/dashboard/prepareQueryRangePayload.ts [hunk 5]
```diff
@@ -91,8 +93,8 @@ export const prepareQueryRangePayload = ({
 	});
 
 	const queryPayload: QueryRangePayload = {
-		start: parseInt(start, 10) * 1e3,
-		end: parseInt(end, 10) * 1e3,
+		start: startTime ? startTime * 1e3 : parseInt(start, 10) * 1e3,
+		end: endTime ? endTime * 1e3 : parseInt(end, 10) * 1e3,
```
Comment (ahmadshaheer) [inline frontend/src/lib/dashboard/prepareQueryRangePayload.ts:97]: can you help me understand, why do we need this?

File: frontend/src/lib/uPlotLib/getUplotChartOptions.ts [hunk 6]
```diff
@@ -222,6 +224,44 @@ export const getUPlotChartOptions = ({
...
+								// const labelText = `Logline: ${label}`;
+								// const labelY = u.bbox.top + 20;
+								// const labelWidth = ctx.measureText(labelText).width + 10;
+								// const labelHeight = 20;
+								// ctx.fillStyle = 'rgba(0, 0, 0, 0.6)';
+								// ctx.fillRect(x - labelWidth / 2, labelY - 15, labelWidth, labelHeight);
+
+								// ctx.fillStyle = 'white';
+								// ctx.fillText(labelText, x, labelY);
+
```
Comment (ahmadshaheer) [inline frontend/src/lib/uPlotLib/getUplotChartOptions.ts:258]: remove commented code?

File: frontend/src/container/LogDetailedView/InfraMetrics/NodeMetrics.tsx (hunk 2 above)
Comment (rahulkeswani101) [reply frontend/src/container/LogDetailedView/InfraMetrics/NodeMetrics.tsx:1]: @ahmadshaheer Initially I kept this as a common component but in node metrics we will change query payloads based on the presence of attribute K8s.node.name and thus created separate components.

File: frontend/src/lib/dashboard/getQueryResults.ts (hunk 4 above)
Comment (rahulkeswani101) [reply frontend/src/lib/dashboard/getQueryResults.ts:91]: Yes, We have start and end time which we use to calculate using prepareQueryRangePayload which further calls getStartEndRangeTime. This depends on what you pass for selectedTime and globalSelectedInterval. But in this case, I have to pass the start and end time based on the log line timestamp.

File: frontend/src/lib/dashboard/prepareQueryRangePayload.ts (hunk 5 above)
Comment (rahulkeswani101) [reply frontend/src/lib/dashboard/prepareQueryRangePayload.ts:97]: Sure @ahmadshaheer. Let's assume someone clicked on some log line which is the time as 2:30 PM then the start time will be 11:30 AM and the end time will be 5:30 PM.  We will subtract 3 hours for start time and add 3 hours for end time based on log line time.

**Review by ahmadshaheer — DISMISSED** [review]

File: frontend/src/container/LogDetailedView/InfraMetrics/NodeMetrics.tsx (hunk 2 above)
Comment (ahmadshaheer) [reply frontend/src/container/LogDetailedView/InfraMetrics/NodeMetrics.tsx:1]: @rahulkeswani101, are we going to build a reusable component for NodeMetrics and PodMetrics in a future PR?

Comment (rahulkeswani101) [reply frontend/src/container/LogDetailedView/InfraMetrics/NodeMetrics.tsx:1]: Yes, @ahmadshaheer. I  have added a comment as a todo for this change in the InfraMetrics component.

**Comment by ankitnayan** [comment]
@rahulkeswani101 please get this released soon.

File: frontend/src/components/LogDetail/index.tsx [hunk 7]
```diff
@@ -246,6 +259,15 @@ function LogDetail({
 					isEdit={isEdit}
 				/>
 			)}
+			{selectedView === VIEW_TYPES.INFRAMETRICS && (
+				<InfraMetrics
+					clusterName={log.resources_string?.[RESOURCE_KEYS.CLUSTER_NAME] as string}
+					podName={log.resources_string?.[RESOURCE_KEYS.POD_NAME] as string}
```
Comment (vikrantgupta25) [inline frontend/src/components/LogDetail/index.tsx:265]: rather than type assertions pass the default as `""` here ?

Comment (vikrantgupta25) [reply frontend/src/components/LogDetail/index.tsx:265]: or make them optional in the props for `<InfraMetrics/>`

File: frontend/src/container/LogDetailedView/InfraMetrics/InfraMetrics.tsx [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+function InfraMetrics({
+	podName,
+	nodeName,
+	hostName,
+	clusterName,
+	logLineTimestamp,
+}: MetricsDataProps): JSX.Element {
+	const initialView = podName ? VIEW_TYPES.POD : VIEW_TYPES.NODE;
```
Comment (vikrantgupta25) [inline frontend/src/container/LogDetailedView/InfraMetrics/InfraMetrics.tsx:27]: if only used as the initial state for selected view then move it inside the useState hook

File: frontend/src/container/LogDetailedView/InfraMetrics/InfraMetrics.tsx [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+			{/* TODO(Rahul): Make a common config driven component for this and other infra metrics components */}
+			{selectedView === VIEW_TYPES.NODE && (
+				<NodeMetrics
+					nodeName={nodeName}
+					clusterName={clusterName}
+					hostName={hostName}
+					logLineTimestamp={logLineTimestamp}
+				/>
+			)}
+			{selectedView === VIEW_TYPES.POD && podName && (
```
Comment (vikrantgupta25) [inline frontend/src/container/LogDetailedView/InfraMetrics/InfraMetrics.tsx:82]: if `podName` is not present there is already an early return in line 34 , why do we need this here ?

File: frontend/src/container/LogDetailedView/InfraMetrics/NodeMetrics.tsx [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+		return {
+			start: startTime.unix(),
+			end: endTime.unix(),
+			verticalLineTimestamp: logTimestamp.unix(),
+		};
+	}, [logLineTimestamp]);
+
+	const queryPayloads = useMemo(() => {
+		if (nodeName) {
```
Comment (vikrantgupta25) [inline frontend/src/container/LogDetailedView/InfraMetrics/NodeMetrics.tsx:51]: nodeName will always be present here. you are early returning in the parent component. the `getHostQueryPayload` will never be accessible. Is this expected ?

File: frontend/src/container/LogDetailedView/InfraMetrics/InfraMetrics.tsx [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+	logLineTimestamp,
+}: MetricsDataProps): JSX.Element {
+	const initialView = podName ? VIEW_TYPES.POD : VIEW_TYPES.NODE;
+	const [selectedView, setSelectedView] = useState<string>(initialView);
+
+	const handleModeChange = (e: RadioChangeEvent): void => {
+		setSelectedView(e.target.value);
+	};
+
+	if (!podName && !nodeName && !hostName) {
```
[the remark this answers is not in the document]
Comment (rahulkeswani101) [reply frontend/src/container/LogDetailedView/InfraMetrics/InfraMetrics.tsx:34]: clusterName is required in query payloads for Node and Pod metrics but not for showing the tabs.

**Review by vikrantgupta25 — APPROVED** [review]

**Review by YounixM — APPROVED** [review]

**Review by YounixM — APPROVED** [review]

**Pull request merged**